README.md

```diff
@@ -0,0 +1,53 @@
+# Pre-initialize WebAssembly Components
+
+[Wizer][] demonstrated that WebAssembly modules can be made to start up
+faster by pre-initializing them. `component-init` does the same thing,
+but for the WebAssembly Component Model.
+
+[Wizer]: https://github.com/bytecodealliance/wizer/
+
+`component-init` is a Rust library that can be used in other projects.
+(One prominent example is [componentize-py][].) To use it in your
+project, you additionally need a WebAssembly runtime that supports the
+Component Model; [Wasmtime][] is known to work.
+
+[componentize-py]: https://github.com/bytecodealliance/componentize-py
+[Wasmtime]: https://wasmtime.dev/
+
+# Design rationale
+
+To pre-initialize a WebAssembly program, you need to do three things:
+
+1. Run its initialization function.
+2. Snapshot any resulting changes to its internal mutable state.
+3. Emit a new WebAssembly program with the updated mutable state.
+
+Step 1 is easy and any runtime will do. Step 3 is a little tedious but
+it just involves copying the original program mostly unchanged except
+for updating data sections and such, so it's not too hard.
+
+The challenge is in step 2: How do you refer to the internal state from
+outside the program? Some state is explicitly exported from a module or
+component, so that's easy to access via any runtime, but what about the
+rest?
+
+To solve this, both Wizer and component-init insert an extra step at the
+beginning: Generate a new WebAssembly program that exports every piece
+of mutable state (e.g. memories and globals) using new names, and
```

Comment:
In the case of `component-init` at least, a component can't export memories or globals, so we do the next best thing: export component-level functions to serve as accessors for those things.  For globals, that means returning whatever scalar component type matches the core wasm time of the global.  For memories, that means returning a `list<u8>`, the contents of which span the entire memory.

Reply:
Ah, clever! I've added a bit more detail to clarify that.